src/internal/operators/share.ts

```diff
@@ -84,33 +101,75 @@ export function share<T>(options: ShareConfig<T>): MonoTypeOperatorFunction<T>;
  * // ... and so on
  * ```
  *
+ * ## Example with notifier factory: Delayed reset
+ * ```ts
+ * import { interval } from 'rxjs';
+ * import { share, take, timer } from 'rxjs/operators';
+ *
+ * const source = interval(1000).pipe(take(3), share({ resetOnRefCountZero: () => timer(1000) }));
+ *
+ * const subscriptionOne = source.subscribe(x => console.log('subscription 1: ', x));
+ * setTimeout(() => subscriptionOne.unsubscribe(), 1300);
+ *
+ * setTimeout(() => source.subscribe(x => console.log('subscription 2: ', x)), 1700);
+ *
+ * setTimeout(() => source.subscribe(x => console.log('subscription 3: ', x)), 5000);
+ *
+ * // Logs:
+ * // subscription 1:  0
+ * // (subscription 1 unsubscribes here)
+ * // (subscription 2 subscribes here ~400ms later, source was not reset)
+ * // subscription 2:  1
+ * // subscription 2:  2
+ * // (subscription 2 unsubscribes here)
+ * // (subscription 3 subscribes here ~2000ms later, source did reset before)
+ * // subscription 3:  0
+ * // subscription 3:  1
+ * // subscription 3:  2
+ * ```
+ *
  * @see {@link api/index/function/interval}
  * @see {@link map}
  *
  * @return A function that returns an Observable that mirrors the source.
  */
-export function share<T>(options?: ShareConfig<T>): OperatorFunction<T, T> {
-  options = options || {};
-  const { connector = () => new Subject<T>(), resetOnComplete = true, resetOnError = true, resetOnRefCountZero = true } = options;
+export function share<T>(options: ShareConfig<T> = {}): MonoTypeOperatorFunction<T> {
+  const { connector = () => new Subject<T>(), resetOnError = true, resetOnComplete = true, resetOnRefCountZero = true } = options;
 
   let connection: SafeSubscriber<T> | null = null;
+  let resetConnection: Subscription | null = null;
   let subject: SubjectLike<T> | null = null;
   let refCount = 0;
   let hasCompleted = false;
   let hasErrored = false;
 
+  const cancelReset = () => {
+    resetConnection?.unsubscribe();
+    resetConnection = null;
+  };
   // Used to reset the internal state to a "cold"
   // state, as though it had never been subscribed to.
   const reset = () => {
+    cancelReset();
     connection = subject = null;
     hasCompleted = hasErrored = false;
   };
+  const resetAndUnsubscribe = () => {
+    // We need to capture the connection before
+    // we reset (if we need to reset).
+    const conn = connection;
+    reset();
+    conn?.unsubscribe();
+  };
 
   return operate((source, subscriber) => {
     refCount++;
+    if (!hasErrored && !hasCompleted) {
+      cancelReset();
+    }
 
     // Create the subject if we don't have one yet.
-    subject = subject ?? connector();
+    const dest = (subject = subject ?? connector());
```

Comment:
The refactor that avoids the non-null assertions is fine, but I'd like to see a brief comment that explains why we have two variables here: `dest` and `subject`. I think it's warranted, as it's not immediately obvious. The comment just needs to point out that `dest` is needed for the notification after `subject` is (potentially) reset.

Reply:
done